lib/CPython.chs

```diff
@@ -164,6 +173,45 @@ endInterpreter :: ThreadState -> IO ()
 endInterpreter (ThreadState ptr) =
   {# call Py_EndInterpreter as ^ #} $ castPtr ptr
 
+newtype GilState = GilState CInt
+-- | Ensure that the current thread is ready to call the Python C API regardless
+-- of the current state of Python, or of the global interpreter lock. This may
+-- be called as many times as desired by a thread as long as each call is matched
+-- with a 'gilStateRelease'. In general, other thread-related APIs may be used
+-- between 'gilStateEnsure' and gilStateRelease calls as long as the thread state
+-- is restored to its previous state before the Release. For this to work
+-- the thread must be a bound thread. (consider using `withGilStateEnsure`
+-- method)
+--
+-- Note: Calling this function from a thread when the runtime is finalizing
+-- will terminate the thread.
+gilStateEnsure :: IO GilState
+gilStateEnsure = do
+  state <- {# call PyGILState_Ensure as ^ #}
+  return $ GilState state
+
+-- | Release any resources previously acquired. After this call,
+-- Python’s -- state will be the same as it was prior to the corresponding
+-- 'gilStateEnsure' call (but generally this state will be unknown to
+-- the caller, hence the use of the GILState API). Every call to
+-- 'gilStateEnsure' must be matched by a call to 'gilStateRelease'
+-- on the same bound thread.
+gilStateRelease :: GilState -> IO ()
+gilStateRelease (GilState s) =
+  {# call PyGILState_Release as ^ #} s
+
+-- | Ensure that the current thread is ready to call the Python C API
+-- regardless of the current state of Python, or of the global interpreter
+-- lock. This method is generally safest way to make python calls in
+-- a multi-threaded scenario, however if you calling this from not bound
+-- threads, it may have some performance implications.
+withGilStateEnsure :: IO a -> IO a
+withGilStateEnsure ioOperation = runInBoundThread $ do
+  gstate <- gilStateEnsure
+  result <- ioOperation
+  gilStateRelease gstate
+  return result
```

Comment:
Might want to use [`bracket`](https://hackage.haskell.org/package/base-4.14.1.0/docs/src/Control.Exception.Base.html#bracket) here

Reply:
:+1: